Mathlib/Analysis/SpecialFunctions/BinEntropy.lean

```diff
@@ -0,0 +1,748 @@
+  by_contra h
+  simp_all
+  obtain ⟨is_diff, hx⟩ := h
+  rw [binaryEntropy_eq] at is_diff
+  by_cases is_x0 : x ≠ 0
+  · rw[hx is_x0] at is_diff
+    apply not_DifferentiableAt_of_DifferentiableAt_add_not_DifferentiableAt _ _ is_diff
+    · apply DifferentiableAt.mul
+       (DifferentiableAt.neg differentiableAt_id') (DifferentiableAt.log differentiableAt_id' _)
+      simp only [ne_eq, one_ne_zero, not_false_eq_true]
+    · have diff_of_aff := (differentiableAt_iff_differentiableAt_comp_mul_add
+        (a:=(0:ℝ)) (b:=(1 : ℝ)) (m:=(-1:ℝ)) (show (-1 : ℝ) ≠ 0 by norm_num) negMulLog).mpr
+      intro h
+      have : DifferentiableAt ℝ (fun (x : ℝ) ↦ (-1 * x + 1).negMulLog) ((-1)⁻¹ * (0 - 1)) := by
+        convert h using 1
+        · ext
+          simp [negMulLog]
+          ring_nf
+        · ring
+      have := diff_of_aff this
+      have := differentiableAt_negMulLog_iff.mp this
+      contradiction
+  · have : x = 0 := by simp_all only [neg_mul, false_implies, ne_eq, Decidable.not_not]
+    rw [this] at is_diff
+    apply not_DifferentiableAt_of_not_DifferentiableAt_add_DifferentiableAt_ _ _ is_diff
+    · simp only [neg_mul, differentiableAt_neg_iff, not_DifferentiableAt_log_mul_zero,
+        not_false_eq_true]
+    · apply DifferentiableAt.neg
+      apply DifferentiableAt.mul
+      apply differentiable_const_minus
+      apply DifferentiableAt.log
+      exact differentiable_const_minus 0
+      simp only [sub_zero, ne_eq, one_ne_zero, not_false_eq_true]
+
+open Filter Real Topology Asymptotics
+
+lemma deriv_log_one_sub_at_1 : deriv (fun p ↦ log (1 - p)) 1 = 0 := by
+  have : ¬ DifferentiableAt ℝ (fun p ↦ log (1 - p)) 1 := by
+    by_contra
+    have : ¬ DifferentiableAt ℝ log 0 := by
+      simp_all only [differentiable_const_minus, implies_true, differentiableAt_log_iff, ne_eq,
+        not_true_eq_false, not_false_eq_true]
+    have : DifferentiableAt ℝ log 0 := by
+      have : Real.log = (fun p ↦ log (1 - p)) ∘ (fun x => 1 - x) := by
+        ext
+        simp only [Function.comp_apply, sub_sub_cancel]
+      rw [this]
+      apply DifferentiableAt.comp
+      · simp only [sub_zero, differentiable_const_minus]
+        assumption
+      · exact differentiable_const_minus 0
+    contradiction
+  exact deriv_zero_of_not_differentiableAt this
+
+lemma deriv_log_one_sub {x : ℝ} : deriv (fun p ↦ log (1 - p)) x = -(1-x)⁻¹ := by
+  by_cases xis1 : x = 1
+  · rw [xis1]
+    simp only [sub_self, inv_zero, neg_zero]
+    exact deriv_log_one_sub_at_1
+  · rw [deriv.log]
+    simp only [deriv_one_minus]
+    field_simp
+    exact differentiable_const_minus x
+    exact sub_ne_zero_of_ne fun a ↦ xis1 a.symm
+
+@[simp] lemma differentiableAt_log_const_neg {x c : ℝ} (h : x ≠ c) :
+    DifferentiableAt ℝ (fun p ↦ log (c - p)) x := by
+  apply DifferentiableAt.log
+  apply DifferentiableAt.sub
+  apply differentiableAt_const
+  apply differentiableAt_id'
+  exact sub_ne_zero.mpr h.symm
+
```

Comment:
```suggestion
@[simp] lemma differentiableAt_log_const_neg {x c : ℝ} (h : x ≠ c) :
    DifferentiableAt ℝ (fun p ↦ log (c - p)) x :=
DifferentiableAt.log (by fun_prop) (sub_ne_zero.mpr h.symm)

```
This can probably be inlined wherever it's used.